src/proto/message_translation/power_frame_msg.cpp

```diff
@@ -0,0 +1,161 @@
+#include <pb_encode.h>
+#include <proto/primitive.nanopb.h>
+
+#ifdef PLATFORMIO_BUILD
+#include <power_frame_msg_platformio.h>
+#else
+#include "pb_decode.h"
+#include "pb_encode.h"
+#include "proto/message_translation/power_frame_msg.h"
+
+#endif
+
+std::vector<uint8_t> serializeToVector(const TbotsProto_PowerFrame& data)
+{
+    std::vector<uint8_t> buffer(TbotsProto_PowerFrame_size);
+    pb_ostream_t stream =
+        pb_ostream_from_buffer(static_cast<uint8_t*>(buffer.data()), buffer.size());
+    if (!pb_encode(&stream, TbotsProto_PowerFrame_fields, &data))
+    {
+        throw std::runtime_error(
+            "Failed to encode PowerFrame msg to vector when converting nanopb to vector");
+    }
+    return buffer;
+}
+
+std::vector<uint8_t> serializeToVector(const TbotsProto_PowerControl& data)
+{
+    std::vector<uint8_t> buffer(TbotsProto_PowerControl_size);
+    pb_ostream_t stream =
+        pb_ostream_from_buffer(static_cast<uint8_t*>(buffer.data()), buffer.size());
+    if (!pb_encode(&stream, TbotsProto_PowerControl_fields, &data))
+    {
+        throw std::runtime_error(
+            "Failed to encode PowerControl msg to vector when converting nanopb to vector");
+    }
+    return buffer;
+}
+
+std::vector<uint8_t> serializeToVector(const TbotsProto_PowerStatus& data)
+{
+    std::vector<uint8_t> buffer(TbotsProto_PowerStatus_size);
+    pb_ostream_t stream =
+        pb_ostream_from_buffer(static_cast<uint8_t*>(buffer.data()), buffer.size());
+    if (!pb_encode(&stream, TbotsProto_PowerStatus_fields, &data))
+    {
+        throw std::runtime_error(
+            "Failed to encode PowerStatus msg to vector when converting nanopb to vector");
+    }
+    return buffer;
+}
```

Comment:
You can template this on the protobuf type and for the fields, you can just take the fields as argument

pb_field_t is the type I think for the TbotsProto_*_fields stuff

Reply:
This would lead to 2 definitions of getMarshalledSize and createUartFrame in uart_framing as there's no way to get the field type from the message. The alternative I thought of was using checking TYPENAME(type) and deciding the pb_field_t that way

Reply:
TYPENAME isnt supported due to not having a rtti lib instead we can use std::is_same